Upgrade broadcast function for multi-dimension variable expansion

adrivinca opened the pull request:
The PR upgrades the `broadcast` function to allow it to work with multiple dimensions correlated in a dataframe

What it was doing before: expand dataframes with an additional dimension
what it does with the update: expand a dataframe with another dataframe with multiple dimensions

example, where the function it is used in combination with the function `pipe()` to pass arguments:

```python
import pandas as pd
from message_ix import make_df
from message_ix_models.util import broadcast

inp0 = make_df(
        "input",
            technology="irrigation_oilcrops",
            value=1,
            unit="-",
            level="water_supply",
            commodity="freshwater",
            mode="M1",
            time="year",
            time_origin="year",
            node_origin=['AFR','NAM'],
    )

# often used to fill one additional variable
inp1 = inp0.pipe(broadcast, year_vtg=[2010,2020])
# it can also be used to add multiple variables distinctly
inp2 = inp0.pipe(broadcast, year_vtg=[2010,2020], 
           year_act=[2010,2020])

# but in this case the variables in inp2 should be mapped
map_v_a = pd.DataFrame({
    'year_vtg' : [2010,2010,2020],
    'year_act' : [2010,2020,2020] }).astype(object)

# this upgrade allows to expand inp0 with the new dataframe map_v_a
inp3 = inp0.pipe(broadcast, map_v_a)
```
I adopted the suggestions from @khaeru in #73 , but I could not get rid of the if block, otherwise I would get error when *args is empty

## How to review

Check if it works, spot cases that might not be covered or that need warning/error messages.
Code improvement/edit suggestions

## PR checklist

  - ~Add or expand tests.~ No change in behaviour, simply refactoring.
  -->
- [ ] Add or expand tests; coverage checks both ✅
- [ ] Add, expand, or update documentation.
- [ ] Update doc/whatsnew.
  <!--
  To do this, add a single line at the TOP of the “Next release” section of
  doc/whatsnew.rst, where '999' is the GitHub pull request number:

  - Title or single-sentence description from above (:pull:`999`:).

  Commit with a message like “Add #999 to doc/whatsnew”
  -->

khaeru commented on message_ix_models/util/__init__.py:
```diff
@@ -161,14 +161,29 @@ def aggregate_codes(df: pd.DataFrame, dim: str, codes):  # pragma: no cover
 
 
-def broadcast(df, **kwargs):
+def broadcast(df, *args, **kwargs):
```
Solid start! Some suggestions:

```python
def broadcast(df, labels: Optional[pd.DataFrame] = None, **kwargs):
```

khaeru commented on message_ix_models/util/__init__.py:
```diff
@@ -161,14 +161,29 @@ def aggregate_codes(df: pd.DataFrame, dim: str, codes):  # pragma: no cover
+                # add dummy common variable between df and labels
+        df['tmp'] = 1
+        labels['tmp'] = 1
+        df = pd.merge(df,labels).drop(columns ={'tmp'})
```
inplace=True in pandas can be expensive/slow, so perhaps something like this:

```python
dfs = [ ]  # Will be a list of pd.DataFrame

# iterrows() produces tuples of (row index, row as pd.Series); use "_" to throw away the index, which is unused
for _, row in labels.iterrows():

    dfs.append(
        # row is like pd.Series({"year_vtg": 2010, "year_act": 2020}) so we can pass it to assign() as keyword args
        df.assign(**row)
    )

# Recombine to a single pd.DataFrame
df = pd.concat(dfs, ignore_index=True)
```

Or using a list comprehension (inline loop):
```python
df = pd.concat([df.assign(**row) for _, row in labels.iterrows()], ignore_index=True)
```

khaeru commented:
> - Add or expand tests

The test to be expanded are here: https://github.com/iiasa/message-ix-models/blob/25ffd8b5adae414690fbd5266df19f27f1d379a8/message_ix_models/tests/test_util.py#L50-L57

adrivinca commented on message_ix_models/util/__init__.py:
```diff
@@ -161,14 +161,29 @@ def aggregate_codes(df: pd.DataFrame, dim: str, codes):  # pragma: no cover
+    labels = pd.DataFrame(*args)
+    if not labels.empty:
+        for cc in labels.columns:
+            if cc in df.columns:
```
thanks! I am incorporating these changes.
Actually, I wanted the function to also work in case the column/dimensions in `labels` are not already present in `df` , i.e. cases where not necessarily `make_df()` is used.
following the example above, I wanted the function to work also for 

```python
inpN = inp0.drop(columns={'year_vtg','year_act'})
# this part of the function is not summorting this case
inp1 = inpN.pipe(broadcast, year_vtg=[2010,2020], 
           year_act=[2010,2020])
# this would work with the if condition
inp2 = inpN.pipe(broadcast, map_v_a)
```
Maybe it is too much, but it would make it more generic and multi-purposes

khaeru replied:
Is this the same `inp0` from the PR description, above? If so, you could do this without drop():
```python
inpN = inp0.assign(year_vtg=None, year_act=None) 
inp1 = inpN.pipe(broadcast, …)
```

This produces the same result, but now the "strict" behaviour of broadcast() can help to detect errors, e.g. if there are some unintended extra dimensions in an argument like `map_v_a`.

adrivinca replied:
ok, agree

adrivinca commented:
Thanks for the great suggestions, I think I followed all of them.
Also added a part to the test, I am not sure if it is the way to write it though

adrivinca commented:
@khaeru can we merge it? I included all your suggestions and it seems to have green light. thanks!

khaeru commented:
> @khaeru can we merge it? I included all your suggestions and it seems to have green light. thanks!

Some things I notice here:
- No one is assigned.
- The PR checklist is not complete. Please ensure the items described are done.
- We probably need more extensive tests that ensure the combo of the new/old functionality works in more elaborate situations.

adrivinca commented:
Thanks @khaeru ! I added a basic test for the old functionality. Not sure if more complex cases are needed. But please let me know if you have any in mind.
Also updated the `whatsnew.rst` and update list above

khaeru commented:
@adrivinca since this PR was opened by you, GitHub doesn't let me request a review from you. But can you look this over (the tests especially) and confirm it will meet your needs?

Some things I did:
- Expanded the tests to cover all possible usages of the function, including incorrect usages that should produce errors.
- Wrote a complete docstring for the function, including an example, and checked that it looked right in the built docs, including the cross-links to other docs.
- Fixed doc/whatsnew. Notice that there was an existing "Next release" line, commented out. Since we don't know what the next release version/date will be until we choose to do that, we just uncomment this heading instead of adding e.g. "2022.5.19".

Total time about 90 minutes.

adrivinca commented:
Thanks @khaeru for your time. I went through the test improvements and it looks great.
Also thanks for fixing the whatsnew, I misinterpreted it.
are we ok to merge now?
